quinn/src/endpoint.rs

```diff
@@ -252,35 +252,36 @@ impl EndpointInner {
     }
 
     fn drive_send(&mut self, cx: &mut Context) -> Result<bool, io::Error> {
-        let mut sent = 0;
-        while let Some(t) = self
-            .outgoing
-            .pop_front()
-            .or_else(|| self.inner.poll_transmit())
-        {
-            match self
-                .socket
-                .poll_send(cx, &t.destination, t.ecn, &t.contents)
-            {
-                Poll::Ready(Ok(_)) => {}
+        let mut calls = 0;
+        loop {
+            while self.outgoing.len() < crate::udp::BATCH_SIZE {
+                match self.inner.poll_transmit() {
+                    Some(x) => self.outgoing.push_back(x),
+                    None => break,
+                }
+            }
+            if self.outgoing.is_empty() {
+                return Ok(false);
+            }
+            match self.socket.poll_send(cx, self.outgoing.as_slices().0) {
```

Comment:
Are you sure this is correct? What invariant do we have that makes sure all the transmits are in the first slice?

Reply:
They don't need to be. `as_slices` returns the slices in order, i.e. `.0` is the first one, and we use `drain` to drop the first `n` that were actually sent. In the common case where the whole batch is sent this never leads to wrapping, but when it does everything still works fine.